*: reduce tracing events during streaming JSON objects

fuweid opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind cleanup


#### What this PR does / why we need it:

If apiserver is handing heavy traffic volume, it's likely to trigger  tracing events. 
After streaming JSON objects, the number of tracing events are same to object number. 
It's unneccessary to log each write call . This patch is to reduce tracing events.
It can reduce memory allocation when the list is to get more than 10,000 objects.


#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes #

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note

```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

fuweid commented:
ping @serathius @liggitt

serathius commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:
```diff
@@ -106,9 +105,24 @@ func SerializeObject(mediaType string, encoder runtime.Encoder, hw http.Response
+			rErr = w.lastErr.Error()
+		}
+
+		span.AddEvent("ResponseWriter",
```
Can you preserve those events ""About to start writing response" and "Write call failed/succeeded" instead of changing the names totally.

serathius commented:
/assign @liggitt

fuweid committed:
*: reduce tracing events during streaming JSON objects

If apiserver is handing heavy traffic volume, it's likely to trigger
tracing events. After streaming JSON objects, the number of tracing events
are same to object number. It's unneccessary to log each write call.
This patch is to reduce tracing events.

Signed-off-by: Wei Fu <fuweid89@gmail.com>

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:
```diff
@@ -282,12 +272,34 @@ func (w *deferredResponseWriter) unbufferedWrite(p []byte) (n int, err error) {
 
+	span := tracing.SpanFromContext(w.ctx)
+	span.AddEvent("About to start writing response",
+		attribute.String("writer", fmt.Sprintf("%T", w.w)),
```
I'd still probably include the size of the initial write... in non-streaming cases, that's the entire object

`attribute.Int("size", len(p))`

in streaming cases, we'll see the difference between the initial write and the total at the end, which is also interesting

fuweid replied:
Updated with "size"

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:
```diff
@@ -200,6 +200,8 @@ type deferredResponseWriter struct {
 	hasWritten  bool
 	hw          http.ResponseWriter
 	w           io.Writer
+	totalBytes  int
```
```suggestion
  // totalBytes is the number of bytes written to `w` and does not include buffered bytes
	totalBytes  int
```

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:
```diff
@@ -200,6 +200,8 @@ type deferredResponseWriter struct {
 	hw          http.ResponseWriter
 	w           io.Writer
+	totalBytes  int
+	lastErr     error
```
```suggestion
  // lastWriteErr holds the error result (if any) of the last write attempt to `w`
	lastWriteErr     error
```

liggitt commented on staging/src/k8s.io/apiserver/pkg/endpoints/handlers/responsewriters/writers.go:
```diff
@@ -242,26 +244,14 @@ func (w *deferredResponseWriter) Write(p []byte) (n int, err error) {
-				attribute.String("writer", fmt.Sprintf("%T", w.w)),
-				attribute.Int("size", len(p)),
-				attribute.Bool("firstWrite", firstWrite))
+			w.lastErr = err
```
I guess we're not worried about a `write error → retry → write success` overwriting a non-nil error with nil

fuweid replied:
It would be easier. Updated!

aaron-prindle commented:
/triage accepted

liggitt commented:
/lgtm
/approve
/ok-to-test
/release-note-none

fuweid commented:
@serathius thanks for the sharing. Glad to see it did improvement!
